File: TUAN_2/2.3.4.cpp
#include <iostream>
#include <vector>

using namespace std;

// Hàm Insertion Sort
void insertionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 1; i < n; i++) {
        int key = arr[i];  // Lưu giá trị hiện tại
        int j = i - 1;

        // Di chuyển các phần tử lớn hơn key lên một vị trí
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;  // Đặt key vào vị trí đúng
    }
}

int main() {
    vector<int> arr = {5, 2, 9, 1, 5};  // Khởi tạo mảng
    insertionSort(arr);  // Gọi hàm sắp xếp

    cout << "Ket qua sau Insertion Sort: ";
    for (int x : arr) {
        cout << x << " " << endl;  // In ra mảng đã sắp xếp
    }
    cout << endl;  // In ra dòng mới sau khi in xong
    return 0;
}

File: TUAN_2/2.3.4.jl
function insertion_sort(arr)
    n = length(arr)
    for i in 2:n
        key = arr[i]  # Lưu giá trị hiện tại
        j = i - 1

        # Di chuyển các phần tử lớn hơn key sang phải
        while j >= 1 && arr[j] > key
            arr[j + 1] = arr[j]
            j -= 1
        end

        # Chèn key vào vị trí đúng
        arr[j + 1] = key
    end

    return arr
end

# Test
arr = [5, 2, 9, 1, 5]  # Khởi tạo mảng
insertion_sort(arr)  # Gọi hàm sắp xếp
println("Kết quả sau Insertion Sort: ", arr)  # In ra mảng đã sắp xếp